In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler, MinMaxScaler
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("feature_engineered_CMG_RRV.csv")

In [2]:
features = [
    'Price',
    'Open',
    'High',
    'Low',
    'Vol.',
    'Change %',
    'return_day',
    'return_week',
    'return_month',
    'volatility_day',
    'volatility_week',
    'volatility_month',
    'liquidity_day',
    'liquidity_week',
    'liquidity_month',
    'z_score',
    'SMA_20',
    'SMA_50',
    'SMA_100',
    'WMA_20',
    'WMA_50',
    'WMA_100',
    'EMA26',
    'EMA12',
    'MACD',
    'Signal_Line',
    'RSI_14',
    'BB_upper',
    'BB_lower',
    'StochRSI_14',
    'Lãi/Lỗ ròng trước thuế',
    'Khấu hao TSCĐ',
    'Dự phòng RR tín dụng',
    'Lãi/Lỗ chênh lệch tỷ giá chưa thực hiện',
    'Lãi/Lỗ từ hoạt động đầu tư',
    'Thu nhập lãi',
    'Thu lãi và cổ tức',
    'Lưu chuyển tiền thuần từ HĐKD trước thay đổi VLĐ',
    'Tăng/Giảm các khoản phải thu',
    'Tăng/Giảm hàng tồn kho',
    'Tăng/Giảm các khoản phải trả',
    'Tăng/Giảm chi phí trả trước',
    'Chi phí lãi vay đã trả',
    'Tiền thu nhập doanh nghiệp đã trả',
    'Tiền thu khác từ các hoạt động kinh doanh',
    'Tiền chi khác từ các hoạt động kinh doanh',
    'Lưu chuyển tiền tệ ròng từ các hoạt động SXKD',
    'Mua sắm TSCĐ',
    'Tiền thu được từ thanh lý tài sản cố định',
    'Tiền chi cho vay, mua công cụ nợ của đơn vị khác (đồng)',
    'Tiền thu hồi cho vay, bán lại các công cụ nợ của đơn vị khác (đồng)',
    'Đầu tư vào các doanh nghiệp khác',
    'Tiền thu từ việc bán các khoản đầu tư vào doanh nghiệp khác',
    'Tiền thu cổ tức và lợi nhuận được chia',
    'Lưu chuyển từ hoạt động đầu tư',
    'Tăng vốn cổ phần từ góp vốn và/hoặc phát hành cổ phiếu',
    'Chi trả cho việc mua lại, trả cổ phiếu',
    'Tiền thu được các khoản đi vay',
    'Tiền trả các khoản đi vay',
    'Cổ tức đã trả',
    'Lưu chuyển tiền từ hoạt động tài chính',
    'Lưu chuyển tiền thuần trong kỳ',
    'Tiền và tương đương tiền',
    'Ảnh hưởng của chênh lệch tỷ giá',
    'Tiền và tương đương tiền cuối kỳ',
    'Tăng trưởng doanh thu (%)',
    'Doanh thu (đồng)',
    'Lợi nhuận sau thuế của Cổ đông công ty mẹ (đồng)',
    'Tăng trưởng lợi nhuận (%)',
    'Thu nhập tài chính',
    'Chi phí tiền lãi vay',
    'Doanh thu bán hàng và cung cấp dịch vụ',
    'Các khoản giảm trừ doanh thu',
    'Doanh thu thuần',
    'Giá vốn hàng bán',
    'Lãi gộp',
    'Chi phí tài chính',
    'Lãi/lỗ từ công ty liên doanh',
    'Chi phí bán hàng',
    'Chi phí quản lý DN',
    'Lãi/Lỗ từ hoạt động kinh doanh',
    'Thu nhập khác',
    'Lãi lỗ trong công ty liên doanh, liên kết',
    'Thu nhập/Chi phí khác',
    'Lợi nhuận khác',
    'LN trước thuế',
    'Chi phí thuế TNDN hiện hành',
    'Chi phí thuế TNDN hoãn lại',
    'Lợi nhuận thuần',
    'Cổ đông thiểu số',
    'Cổ đông của Công ty mẹ',
    '(Vay NH+DH)/VCSH',
    'Nợ/VCSH',
    'TSCĐ / Vốn CSH',
    'Vốn CSH/Vốn điều lệ',
    'Vòng quay tài sản',
    'Vòng quay TSCĐ',
    'Số ngày thu tiền bình quân',
    'Số ngày tồn kho bình quân',
    'Số ngày thanh toán bình quân',
    'Chu kỳ tiền',
    'Vòng quay hàng tồn kho',
    'Biên EBIT (%)',
    'Biên lợi nhuận gộp (%)',
    'Biên lợi nhuận ròng (%)',
    'ROE (%)',
    'ROIC (%)',
    'ROA (%)',
    'EBITDA (Tỷ đồng)',
    'EBIT (Tỷ đồng)',
    'Tỷ suất cổ tức (%)',
    'Chỉ số thanh toán hiện thời',
    'Chỉ số thanh toán tiền mặt',
    'Chỉ số thanh toán nhanh',
    'Khả năng chi trả lãi vay',
    'Đòn bẩy tài chính',
    'Vốn hóa (Tỷ đồng)',
    'Số CP lưu hành (Triệu CP)',
    'P/E',
    'P/B',
    'P/S',
    'P/Cash Flow',
    'EPS (VND)',
    'BVPS (VND)',
    'EV/EBITDA'
]


In [3]:
data = df[features].values

# scaler = StandardScaler()
scaler = MinMaxScaler()
data_scaled = scaler.fit_transform(data)

In [4]:
len(features)

125

In [5]:
df.head()

,Date,Price,Open,High,Low,Vol.,Change %,Symbol,return_day,return_week,...,Đòn bẩy tài chính,Vốn hóa (Tỷ đồng),Số CP lưu hành (Triệu CP),P/E,P/B,P/S,P/Cash Flow,EPS (VND),BVPS (VND),EV/EBITDA
0,2013-04-01,3299.1,3181.3,3358.0,3181.3,17230.0,-1.75,CMG_RRV,-0.017540,-0.034476,...,2.897513,291467853600,66242694,24.173205,0.501469,0.104289,4.587862,-22.507076,8774.213429,5.796877
1,2013-04-02,3358.0,3299.1,3358.0,3299.1,56370.0,1.79,CMG_RRV,0.017853,0.000000,...,2.897513,291467853600,66242694,24.173205,0.501469,0.104289,4.587862,-22.507076,8774.213429,5.796877
2,2013-04-03,3358.0,3358.0,3416.9,3240.2,94080.0,0.00,CMG_RRV,0.000000,0.000000,...,2.897513,291467853600,66242694,24.173205,0.501469,0.104289,4.587862,-22.507076,8774.213429,5.796877
3,2013-04-04,3299.1,3358.0,3358.0,3181.3,53550.0,-1.75,CMG_RRV,-0.017540,0.000000,...,2.897513,291467853600,66242694,24.173205,0.501469,0.104289,4.587862,-22.507076,8774.213429,5.796877
4,2013-04-05,3299.1,3299.1,3358.0,3122.3,75900.0,0.00,CMG_RRV,0.000000,-0.017540,...,2.897513,291467853600,66242694,24.173205,0.501469,0.104289,4.587862,-22.507076,8774.213429,5.796877


In [6]:
df.tail()

,Date,Price,Open,High,Low,Vol.,Change %,Symbol,return_day,return_week,...,Đòn bẩy tài chính,Vốn hóa (Tỷ đồng),Số CP lưu hành (Triệu CP),P/E,P/B,P/S,P/Cash Flow,EPS (VND),BVPS (VND),EV/EBITDA
2914,2024-12-25,50360.0,50270.0,50991.0,49910.0,2290000.0,1.63,CMG_RRV,0.016347,0.005390,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2915,2024-12-26,49910.0,50811.0,50811.0,49910.0,1560000.0,-0.89,CMG_RRV,-0.008936,-0.019469,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2916,2024-12-27,48559.0,49730.0,49910.0,48468.0,3600000.0,-2.71,CMG_RRV,-0.027069,-0.046011,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2917,2024-12-30,49100.0,49000.0,49500.0,49000.0,963700.0,1.11,CMG_RRV,0.011141,-0.023274,...,2.227640,9468086700800,211341221,29.761275,3.458975,1.222675,14.522856,278.574989,12951.813926,14.864212
2918,2024-12-31,49250.0,49300.0,49850.0,48950.0,957300.0,0.31,CMG_RRV,0.003055,-0.006054,...,2.152912,8695028208050,211300807,27.838678,3.047378,1.095932,11.593105,599.010040,13503.411529,12.007542


In [7]:
len(df)

2919

In [8]:
window_size = 30
forecast_horizon = 30

def create_sequences(data, window_size, forecast_horizon):
    X, y = [], []
    for i in range(len(data) - window_size - forecast_horizon + 1):
        X.append(data[i : i + window_size])
        y.append(data[i + window_size : i + window_size + forecast_horizon, 0])
    return np.array(X), np.array(y)

X, y = create_sequences(data_scaled, window_size, forecast_horizon)
print("X shape:", X.shape)  # Kích thước: (số mẫu, window_size, số đặc trưng)
print("y shape:", y.shape)  # Kích thước: (số mẫu, forecast_horizon)

X shape: (2860, 30, 125)
y shape: (2860, 30)


With 125 features, you may have redundant or noisy information. It can be beneficial to first explore feature selection or dimensionality reduction (e.g., PCA) rather than simply increasing model complexity.

In [9]:
model = Sequential()
model.add(LSTM(100, activation='relu', return_sequences=True, input_shape=(window_size, X.shape[2])))
model.add(Dropout(0.2))  

model.add(LSTM(50, activation='relu'))
model.add(Dropout(0.2))  
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.2))  
model.add(Dense(forecast_horizon))
model.compile(optimizer='adam', loss='mse')
model.summary()

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/keras/src/layers/rnn/rnn.py:200: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ lstm (LSTM)                     │ (None, 30, 100)        │        90,400 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 30, 100)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_1 (LSTM)                   │ (None, 50)             │        30,200 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 50)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 32)             │         1,632 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 32)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 30)             │           990 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 123,222 (481.34 KB)

 Trainable params: 123,222 (481.34 KB)

 Non-trainable params: 0 (0.00 B)

In [10]:
# model training
history = model.fit(X, y, epochs=20, batch_size=64, validation_split=0.2)

Epoch 1/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 2s 30ms/step - loss: 0.1056 - val_loss: 0.1267
Epoch 2/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 26ms/step - loss: 0.0436 - val_loss: 0.0350
Epoch 3/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - loss: 0.0217 - val_loss: 0.0159
Epoch 4/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - loss: 0.0146 - val_loss: 0.0131
Epoch 5/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 29ms/step - loss: 0.0115 - val_loss: 0.0102
Epoch 6/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - loss: 0.0094 - val_loss: 0.0130
Epoch 7/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - loss: 0.0084 - val_loss: 0.0114
Epoch 8/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 26ms/step - loss: 0.0072 - val_loss: 0.0139
Epoch 9/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - loss: 0.0062 - val_loss: 0.0134
Epoch 10/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - loss: 0.0058 - val_loss: 0.0110
Epoch 11/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 28ms/step - loss: 0.0052 - val_loss: 0.0126
Epoch 12/20
36/36 ━━━━━━━━━━━━━━━━━━━━ 1s 27ms/step - loss: 0.0

In [11]:
# Prediction for the next 30 days
last_window = data_scaled[-window_size:]
last_window = np.expand_dims(last_window, axis=0) 

In [12]:
# Predict using the model
predicted_scaled = model.predict(last_window)  # Shape: (1, forecast_horizon)
predicted_scaled.shape

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 115ms/step


(1, 30)

In [13]:
# Create dummy array to inverse transform (assume 'Price' is at index 0)
dummy = np.zeros((forecast_horizon, len(features)-1))
pred_full = np.concatenate([predicted_scaled.T, dummy], axis=1)
predicted_prices = scaler.inverse_transform(pred_full)[:, 0]  # Extract the 'Price' column

print("Dự báo giá trong {} ngày tới cho mã là:".format(forecast_horizon))
print(predicted_prices)

Dự báo giá trong 30 ngày tới cho mã là:
[47418.48571081 48459.62251921 46978.48172617 47204.49013605
 47958.17643766 48294.28295784 46706.53649864 48352.80063667
 47233.48145351 47526.21643505 48278.42220421 47050.32786179
 46968.0447053  48423.408109   48291.86060152 49121.80531864
 46778.05647736 47777.26196804 47485.28924084 47484.74686756
 48278.64941463 47610.35392179 48593.58138895 48292.72546701
 47840.41180782 47494.83574343 47777.668748   47323.02068491
 47819.69901218 48295.56926203]


In [14]:
daily_file = "processed_stck_data/CMG_RRV.csv"
df_daily = pd.read_csv(daily_file, parse_dates=['Date'])
df_daily.sort_values("Date", inplace=True)
last_date_str = "2024-12-31"
last_date = pd.to_datetime(last_date_str)

In [15]:
# Training cutoff was 31/12/2024, so the forecast period is 01/01/2025 -> 31/01/2025
forecast_window_start = pd.to_datetime("2025-01-01")
forecast_window_end   = pd.to_datetime("2025-01-31")

# Filter the daily data to only include the forecast period
mask = (df_daily['Date'] >= forecast_window_start) & (df_daily['Date'] <= forecast_window_end)
df_forecast_real = df_daily.loc[mask]

# Extract the 'Price' values (these are the real prices for the forecast period)
real_prices_next_30 = df_forecast_real['Price'].values

print("Real Price values for 01/01/2025 -> 31/01/2025:")
print(real_prices_next_30)

Real Price values for 01/01/2025 -> 31/01/2025:
[49300. 49350. 49000. 48150. 47200. 46000. 44650. 44800. 44800. 45100.
 46100. 47450. 46600. 46950. 46400. 47500. 47100.]
